File: daily_challenges/1733.py


# Tc = O(F * L)
from collections import defaultdict

class Solution:
    def minimumTeachings(self, n: int, languages: list[list[int]], friendships: list[list[int]]) -> int:
        sad_users = set()
        
        for friends in friendships:
            u = friends[0] - 1  # convert this to 0 base indexing
            v = friends[1] - 1

            lang_set = set(languages[u])

            can_talk = False
            for lang in languages[v]:
                if lang in lang_set:
                    can_talk = True
                    break
            if not can_talk:
                sad_users.add(u)
                sad_users.add(v)

        # find max known language among sad users
        language = [0] * (n + 1)
        most_known_lang = 0
        for user in sad_users:
            for lang in languages[user]:
                language[lang] += 1
                most_known_lang = max(most_known_lang, language[lang])

        return len(sad_users) - most_known_lang

File: daily_challenge_leetcode/1733.js
/**
 * @param {number} n
 * @param {number[][]} languages
 * @param {number[][]} friendships
 * @return {number}
 */
var minimumTeachings = function (n, languages, friendships) {
    const sadUsers = new Set();

    for (const friends of friendships) {
        const u = friends[0] - 1; 
        const v = friends[1] - 1;

        const langSet = new Set(languages[u]);

        let canTalk = false;
        for (const lang of languages[v]) {
            if (langSet.has(lang)) {
                canTalk = true;
                break;
            }
        }
        if (!canTalk) {
            sadUsers.add(u);
            sadUsers.add(v);
        }
    }

   
    const language = new Array(n + 1).fill(0);
    let mostKnownLang = 0;
    for (const user of sadUsers) {
        for (const lang of languages[user]) {
            language[lang] += 1;
            mostKnownLang = Math.max(mostKnownLang, language[lang]);
        }
    }

    return sadUsers.size - mostKnownLang;

};